044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,296 @@
+stop message production) for all limit types.
+
+The only requirement is a single hard-limit configuration, if no soft-limit is configured we will infer that soft-limit==hard-limit.
+
+The different limit types can be combined, so you could combine a usedBytesExceeds soft limit with a availableRatioBelow
+hard limit.
+
+Defining multiple soft limits or multiple hard limits would be an invalid state.
+
+The limits we want are:
+1. throttle if [usedBytes](#volume) is greater-than-or-equal-to some threshold for any volume
+2. throttle if [availableBytes](#volume) is less-than-or-equal-to some threshold for any volume
+3. throttle if the [availableRatio](#volume) is less-than-or-equal-to some threshold for any volume
+
+Using configuration properties like:
+
+1. `client.quota.callback.static.storage.perVolumeLimit.usedBytesExceeds.soft=14000000000`
+2. `client.quota.callback.static.storage.perVolumeLimit.usedBytesExceeds.hard=15000000000`
+3. `client.quota.callback.static.storage.perVolumeLimit.availableBytesBelow.soft=5000000000`
+4. `client.quota.callback.static.storage.perVolumeLimit.availableBytesBelow.hard=1000000000`
+5. `client.quota.callback.static.storage.perVolumeLimit.availableRatioBelow.soft=0.05`
+6. `client.quota.callback.static.storage.perVolumeLimit.availableRatioBelow.hard=0.01`
+
+`availableRatioBelow` value can be in the range (0.0, 1.0)
+
+#### Fallback Throttle Factor
+
+We are going to depend on using the admin client to get volume information. This brings all the baggage of making a
+connection and dealing with possible failures. Also, we can potentially have an inconsistent view of the world between
+determining the active broker set and asking for a description of all log dirs (because we make two independent API calls)
+. So we need to react somehow to cases where we cannot get the volume data for all active brokers.
+
+Example inconsistent state:
+1. we call `describeCluster` and get a response that says broker 1 and 2 are active
+2. broker 2 shuts down cleanly and is removed from the active set
+3. we call `describeLogDirs( brokerIds = [1,2] )` and only receive descriptions for logdirs on broker 1
+
+We propose introducing a configurable **fallback throttle factor** to be applied in situations where we don't have enough
+information to act. With a default value of 0.0 ie effectively stopping all production to the cluster. This would allow
+users to opt in to more dangerous behaviour like disabling this failsafe.
+
+Fallback throttle factor can be in the range (0.0, 1.0)
+
+Example configuration: `client.quota.callback.static.fallback.throttle.factor=0.0`
+
+## Metrics
+
+1. Add a `io.strimzi.kafka.quotas:type=LocalThrottleFactor,name=ThrottleFactor` gauge. This would emit the most recently calculated throttle factor.
+2. Add a `io.strimzi.kafka.quotas:type=LocalThrottleFactor,name=FallbackThrottleFactorApplied,reason=xyz` counter. Where reason could be something like 'connection error'.
```

Comment:
I'm not quite sure if this is helpful because that could be just a transient error, and it'll be put in metrics forever? And also are there will be many different error metrics? 

Maybe we can add:
1. lastFallbackThrottleError : it'll record the last error we encountered
2. last timestamp we successfully retrieve log dir info: In addition to (1), we also need this metric, so that we can know this error in (1) is happening now, or it's just an old error we can ignore.

Ex: if we saw metrics like this:
metric (1): SSL certificate expiration error
metric (2): 123455667 (suppose it's 30 mins ago)

So that we can know, ok, it's been 30 mins we haven't successfully got the log dir info, and the reason is "SSL certificate expiration error".

In another case
metric (1): request timeout
metric (2): 123455667 (suppose it's 10 sec ago)

So we can know the error should be a transient error and it's working well now. 

Does that make sense?

Reply:
Assuming it's a monotonic counter and getting scraped into something like prometheus, you'd be able to see when the error counter incremented in time.

But yes, if we're thinking of users connecting with JMX the raw error counts wouldn't be so helpful. Maybe counters plus error logging is enough.

Reply:
Or assuming you can at least enumerate the possible errors use a guage whose value is the integer value of an enum (like BrokerState). 0 => Everything is fine. No good for arbitrary messages, but enough for alerting and then the SOP directs an SRE has to look at the logs.

Reply:
Alright to sidestep the issue I've dropped the `reason` label :grin: , this one might need more discovery once something is coded up. For now we could use a counter plus error logging.